### PR TITLE
Add Op(_scaled_dot_product_flash_attention) | feat(torchlib)

### PR DESCRIPTION
**Pull request by titaiwangms**
`_scaled_dot_product_flash_attention` is one out of three ATen implementations of `nn.functional.scaled_dot_product_attention` according to the page: https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html.

As of which one of three ATen operator is representing `nn.functional.scaled_dot_product_attention` in a model is decided by a context manager: https://pytorch.org/docs/stable/backends.html. From ONNX perspective, they have no difference except the function signature.

Only the first result matters in terms of the model prediction, and the unrelated outputs are following the below code:
```python
@register_meta(
    [
        aten._scaled_dot_product_flash_attention,
    ]
)
def meta__scaled_dot_product_flash(
    query: Tensor,
    key: Tensor,
    value: Tensor,
    dropout_p: float = 0.0,
    is_causal: bool = False,
    return_debug_mask: bool = False,
    scale: Optional[float] = None,
):
    batch_size = query.size(0)
    num_heads = query.size(1)
    max_seqlen_batch_q = query.size(2)
    head_dim = query.size(3)

    max_seqlen_batch_k = key.size(2)
    if device_hint(query) == "cpu":
        Nnz_q = batch_size * max_seqlen_batch_q
        query_t = query.transpose(1, 2)
        query_reshaped = query_t.reshape(Nnz_q, num_heads, head_dim)
        attention = torch.empty_like(query_reshaped, device=query.device)
        attention = attention.view(
            batch_size, max_seqlen_batch_q, num_heads, head_dim
        ).transpose(1, 2)
        logsumexp = torch.empty(
            (
                batch_size,
                max_seqlen_batch_q,
                num_heads,
            ),
            dtype=torch.float,
            device=query.device,
        ).transpose(1, 2)
        return (
            attention,
            logsumexp,
            torch.empty((), dtype=torch.int32, device="meta"),
            torch.empty((), dtype=torch.int32, device="meta"),
            0,
            0,
            torch.empty((), dtype=torch.long, device="meta"),
            torch.empty((), dtype=torch.long, device="meta"),
            torch.empty((), dtype=query.dtype, device=query.device),
        )

    # Cuda Path
    query_t = query.transpose(1, 2)
    attention = torch.empty_like(query_t).transpose(1, 2)
    logsumexp = torch.empty(
        (batch_size, num_heads, max_seqlen_batch_q),
        dtype=torch.float,
        device=query.device,
    )
    cumulative_sequence_length_q = torch.empty(
        batch_size + 1, dtype=torch.int32, device="meta"
    )
    cumulative_sequence_length_k = torch.empty(
        batch_size + 1, dtype=torch.int32, device="meta"
    )

    if return_debug_mask:
        blocksize_c = 128 if head_dim > 64 else 256
        max_seqlen_k = math.ceil(max_seqlen_batch_q / blocksize_c)
        if max_seqlen_batch_k <= 128:
            max_seqlen_k = 128
        elif max_seqlen_batch_k <= 256:
            max_seqlen_k = 256
        debug_mask = torch.empty(
            (batch_size, num_heads, max_seqlen_batch_q, max_seqlen_k),
            dtype=query.dtype,
            device=query.device,
        )
    else:
        debug_mask = torch.empty(0, dtype=query.dtype, device=query.device)

    # Note [Seed and Offset]: device for seed and offset below depends on whether we are
    # capturing or not, but at the time of tracing we don't know if we
    # are going to use cudagraphs or not, so we return meta tensors here
    # it's possible we'll need to have some special handling in inductor for sdpa

    return (
        attention,
        logsumexp,
        None,
        None,
        max_seqlen_batch_q,
        max_seqlen_batch_k,
        torch.empty((), dtype=torch.long, device="meta"),
        torch.empty((), dtype=torch.long, device="meta"),
        debug_mask,
    )
```

NOTE: PyTorch converter should consider None would appear in `_fill_tensor_shape_type`, otherwise, the exporter crashes.

### CODE REVIEW

File: onnxscript/function_libs/torch_lib/ops/nn.py
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/nn.py:1700]:
```suggestion
    logsumexp = op.Expand(0.0, op.Shape(query))
```

File: onnxscript/tests/function_libs/torch_lib/ops_test_data.py
Comment (justinchuby) [inline onnxscript/tests/function_libs/torch_lib/ops_test_data.py:1753]:
```suggestion
        "ops.aten._scaled_dot_product_flash_attention",
```

typo?

**Review by justinchuby — APPROVED** [review]

File: onnxscript/function_libs/torch_lib/ops/nn.py
Comment (titaiwangms) [inline onnxscript/function_libs/torch_lib/ops/nn.py:1707]: Blocked by 
```
onnx.onnx_cpp2py_export.checker.ValidationError: Attribute 'value_ints' is expected to have field 'ints'
```

Comment (fatcat-z) [reply onnxscript/function_libs/torch_lib/ops/nn.py:1707]: Is this an onnx issue and you are going to fix it firstly?

Comment (titaiwangms) [reply onnxscript/function_libs/torch_lib/ops/nn.py:1707]: https://github.com/onnx/onnx/pull/5559 There is a PR in ONNX trying to unblock this usage.

Comment (titaiwangms) [reply onnxscript/function_libs/torch_lib/ops/nn.py]: Use `make_tensor` to unblock for now.

File: onnxscript/function_libs/torch_lib/ops/nn.py
Comment (titaiwangms) [inline onnxscript/function_libs/torch_lib/ops/nn.py:1735]: Should I create `TInt` for these guys?

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/nn.py:1735]: INT64?

Comment (titaiwangms) [reply onnxscript/function_libs/torch_lib/ops/nn.py:1735]: The one in embedding remains TFloat though.

**Comment by justinchuby** [comment]
lgtm with the return types fixed

File: onnxscript/function_libs/torch_lib/ops/nn.py
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/nn.py:1735]: If the return types for the empty float values need to be TFloat, do we need a CaskLike self here? Otherwise it would be FLOAT because the dtype is set and not dependent on the input?

**Comment by titaiwangms** [comment]
@justinchuby I found CI all fails except torch-nightly. I guess it needs torch-nightly to test this op?

**Comment by justinchuby** [comment]
> @justinchuby I found CI all fails except torch-nightly. I guess it needs torch-nightly to test this op?

Looks like so. We can skip the tests for older torch by using `.skip(enabled_if=version_utils.torch_older_than("2.1"))`

**Review by justinchuby — APPROVED** [review]

**Pull request merged**
